File: Diagrammatize.Design.Pattern/code/cxx/Flyweight/A1/BigString.cpp
#include "BigString.hpp"
#include "BigCharFactory.hpp"

BigString::BigString(const std::string& string, bool shared) {
    if (shared) {
        initShared(string);
    } else {
        initUnshared(string);
    }
}

void BigString::initShared(const std::string& string) {
    bigchars.resize(string.length());
    auto& factory = BigCharFactory::getInstance();
    for (int i = 0; i < bigchars.size(); i++) {
        bigchars[i] = factory.getBigChar(string[i]);
    }
}

void BigString::initUnshared(const std::string& string) {
    bigchars.resize(string.length());
    for (int i = 0; i < bigchars.size(); i++) {
        bigchars[i] = new BigChar(string[i]);
    }
}

void BigString::print() {
    for (int i = 0; i < bigchars.size(); i++) {
        bigchars[i]->print();
    }
}

File: Diagrammatize.Design.Pattern/code/python/Flyweight/A1/BigString.py

from BigCharFactory import BigCharFactory
from BigChar import BigChar

class BigString:
    def __init__(self, string, shared):
        if shared:
            self.initShared(string)
        else:
            self.initUnshared(string)

    def initShared(self, string):
        bigchars = [None]*len(string)
        factory = BigCharFactory.getInstance()
        for i in range(len(string)):
            bigchars[i] = factory.getBigChar(string[i])
        self.bigchars = bigchars

    def initUnshared(self, string):
        bigchars = [None]*len(string)
        for i in range(len(string)):
            bigchars[i] = BigChar(string[i])
        self.bigchars = bigchars

    def print(self):
        for i in range(len(self.bigchars)):
            self.bigchars[i].print()
